File: Awesome.AI.Source/Awesome.AI/Factorys/BotFactory.cs
using Awesome.AI.Core;
using Awesome.AI.Variables;
using static Awesome.AI.Variables.Enums;

namespace Awesome.AI.Factorys
{
    public interface IBot
    {
        public MINDS mindtype { get; set; }
        public LOGICTYPE logic { get; set; }
        public MECHANICS mech_low { get; set; }
        public MECHANICS mech_high { get; set; }
        public Dictionary<LONGTYPE, string> lng_dec { get; set; }
        public VALIDATION validation { get; set; }
        public TAGS tags { get; set; }                                               //used with WORLD and BOTH
        public OCCUPASION occupasion { get; set; }                                   //used with SELF and BOTH
        public PATTERN pattern { get; set; }
        public ARC arc { get; set; }
        public int RUNTIME { get; set; }                                               //minutes / 2

    }

    public class Roberta : IBot
    {
        public MINDS mindtype { get; set; }
        public LOGICTYPE logic { get; set; }
        public MECHANICS mech_low { get; set; }
        public MECHANICS mech_high { get; set; }
        public Dictionary<LONGTYPE, string> lng_dec { get; set; }
        public VALIDATION validation { get; set; }
        public TAGS tags { get; set; }                                               //used with WORLD and BOTH
        public OCCUPASION occupasion { get; set; }                                   //used with SELF and BOTH
        public PATTERN pattern { get; set; }
        public ARC arc { get; set; }
        public int RUNTIME { get; set; }

        public Roberta()
        {
            mindtype = MINDS.ROBERTA;
            logic = LOGICTYPE.PROBABILITY;
            mech_low = MECHANICS.TUGOFWAR_LOW;
            mech_high = MECHANICS.BALLONHILL_HIGH;
            lng_dec = CONST.lng_dec_roberta;

            validation = VALIDATION.BOTH;                                       //BOTH or TAGS
            tags = TAGS.ALL;                                                    //used with TAGS and BOTH
            occupasion = OCCUPASION.DYNAMIC;                                    //used with OCCU and BOTH
            pattern = PATTERN.MOODGENERAL;
            arc = ARC.DONTUSE;
            RUNTIME = (6) / 2;                                                  //minutes / 2
    }
    }

    public class Andrew : IBot
    {
        public MINDS mindtype { get; set; }
        public LOGICTYPE logic { get; set; }
        public MECHANICS mech_low { get; set; }
        public MECHANICS mech_high { get; set; }
        public Dictionary<LONGTYPE, string> lng_dec { get; set; }
        public VALIDATION validation { get; set; }
        public TAGS tags { get; set; }                                               //used with WORLD and BOTH
        public OCCUPASION occupasion { get; set; }                                   //used with SELF and BOTH
        public PATTERN pattern { get; set; }
        public ARC arc { get; set; }
        public int RUNTIME { get; set; }

        public Andrew()
        {
            mindtype = MINDS.ANDREW;
            logic = LOGICTYPE.PROBABILITY;
            mech_low = MECHANICS.TUGOFWAR_LOW;
            mech_high = MECHANICS.TUGOFWAR_HIGH;
            lng_dec = CONST.lng_dec_andrew;

            validation = VALIDATION.BOTH;                                       //BOTH or TAGS
            tags = TAGS.ALL;                                                    //used with TAGS and BOTH
            occupasion = OCCUPASION.DYNAMIC;                                    //used with OCCU and BOTH
            pattern = PATTERN.MOODGENERAL;
            arc = ARC.DONTUSE;
            RUNTIME = (6) / 2;
        }
    }

    public class Basic : IBot
    {
        public MINDS mindtype { get; set; }
        public LOGICTYPE logic { get; set; }
        public MECHANICS mech_low { get; set; }
        public MECHANICS mech_high { get; set; }
        public Dictionary<LONGTYPE, string> lng_dec { get; set; }
        public VALIDATION validation { get; set; }
        public TAGS tags { get; set; }                                               //used with WORLD and BOTH
        public OCCUPASION occupasion { get; set; }                                   //used with SELF and BOTH
        public PATTERN pattern { get; set; }
        public ARC arc { get; set; }
        public int RUNTIME { get; set; }

        public Basic()
        {
            mindtype = MINDS.BASIC;
            logic = LOGICTYPE.PROBABILITY;
            mech_low = MECHANICS.TUGOFWAR_LOW;
            mech_high = MECHANICS.BALLONHILL_HIGH;
            lng_dec = CONST.lng_dec_basic;

            validation = VALIDATION.BOTH;                                       //BOTH or TAGS
            tags = TAGS.ALL;                                                    //used with TAGS and BOTH
            occupasion = OCCUPASION.DYNAMIC;                                    //used with OCCU and BOTH
            pattern = PATTERN.MOODGENERAL;
            arc = ARC.USE;
            RUNTIME = (120) / 2;                                                  //minutes / 2
        }
    }

    public class BotFactory
    {
        private MINDS mindtype { get; set; }

        public TheMind mind;
        private BotFactory() { }
        public BotFactory(TheMind mind)
        {
            this.mind = mind;
            this.mindtype = mind.mindtype;
        }

        public IBot GetBot()
        {
            IBot bot = null;

            switch (mindtype)
            {
                case MINDS.ROBERTA: bot = new Roberta(); break;
                case MINDS.ANDREW: bot = new Andrew(); break;
                case MINDS.BASIC: bot = new Basic(); break;
                default: throw new Exception("Bots, GetBot");
            }            

            return bot;
        }
    }
}

File: Awesome.AI.Python/Factorys/BotFactory.py
"""Bot definitions and factory ported from ``Factorys/BotFactory.cs``.

ARC is deliberately excluded from the Python port.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from awesome_ai.Variables.Constants import CONST
from awesome_ai.Variables.Enums import (
    LONGTYPE,
    LOGICTYPE,
    MECHANICS,
    MINDS,
    OCCUPASION,
    PATTERN,
    PROPS,
    TAGS,
    VALIDATION,
)

if TYPE_CHECKING:
    from awesome_ai.Core.TheMind import TheMind


@dataclass
class IBot:
    """Python data equivalent of the C# ``IBot`` property contract."""

    mindtype: MINDS
    logic: LOGICTYPE
    mech_low: MECHANICS
    props: PROPS
    lng_dec: dict[LONGTYPE, str]
    validation: VALIDATION
    tags: TAGS
    occupasion: OCCUPASION
    pattern: PATTERN
    RUNTIME: int


class Roberta(IBot):
    def __init__(self) -> None:
        super().__init__(
            MINDS.ROBERTA,
            LOGICTYPE.SHARED,
            MECHANICS.TUGOFWAR_LOW,
            PROPS.TEMPERAMENT,
            CONST.lng_dec_roberta,
            VALIDATION.BOTH,
            TAGS.ALL,
            OCCUPASION.DYNAMIC,
            PATTERN.MOODGENERAL,
            6 // 2,
        )


class Andrew(IBot):
    def __init__(self) -> None:
        super().__init__(
            MINDS.ANDREW,
            LOGICTYPE.SHARED,
            MECHANICS.TUGOFWAR_LOW,
            PROPS.TEMPERAMENT,
            CONST.lng_dec_andrew,
            VALIDATION.BOTH,
            TAGS.ALL,
            OCCUPASION.DYNAMIC,
            PATTERN.MOODGENERAL,
            6 // 2,
        )


class Basic(IBot):
    def __init__(self) -> None:
        super().__init__(
            MINDS.BASIC,
            LOGICTYPE.SHARED,
            MECHANICS.TUGOFWAR_LOW,
            PROPS.TEMPERAMENT,
            CONST.lng_dec_basic,
            VALIDATION.BOTH,
            TAGS.ALL,
            OCCUPASION.DYNAMIC,
            PATTERN.MOODGENERAL,
            120 // 2,
        )


class BotFactory:
    def __init__(self, mind: TheMind) -> None:
        self.mind = mind
        self.mindtype = mind.mindtype

    def GetBot(self) -> IBot:
        if self.mindtype == MINDS.ROBERTA:
            return Roberta()
        if self.mindtype == MINDS.ANDREW:
            return Andrew()
        if self.mindtype == MINDS.BASIC:
            return Basic()
        raise Exception("Bots, GetBot")
